plugins/logitech-tap/fu-logitech-tap-touch-device.c

```diff
@@ -0,0 +1,913 @@
+		return FALSE;
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_logitech_tap_touch_device_check_busy_cb,
+				  TAP_TOUCH_MAX_BUSY_CHECK_RETRY,
+				  5,
+				  NULL,
+				  error)) {
+		g_prefix_error(error, "failed to crc for %s, device busy", in_ap ? "AP" : "DF");
+		return FALSE;
+	}
+	if (!fu_logitech_tap_touch_device_get_crc(self, &device_checksum, 4, error))
+		return FALSE;
+	if (device_checksum != firmware_checksum) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_DATA,
+			    "crc check failed for %s, expected 0x%04x and got 0x%04x",
+			    in_ap ? "AP" : "DF",
+			    firmware_checksum,
+			    device_checksum);
+		return FALSE;
+	}
+	g_info("device checksum for %s. checksum:0x%x", in_ap ? "AP" : "DF", device_checksum);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_touch_device_write_firmware(FuDevice *device,
+					    FuFirmware *firmware,
+					    FuProgress *progress,
+					    FwupdInstallFlags flags,
+					    GError **error)
+{
+	FuLogitechTapTouchDevice *self = FU_LOGITECH_TAP_TOUCH_DEVICE(device);
+	guint32 ap_start = 0x2000;
+	guint32 ap_end = 0;
+	guint16 ap_check = 0;
+	guint32 df_start = 0xF000;
+	guint32 df_end = fu_firmware_get_size(firmware) - 1;
+	guint32 df_check = 0;
+	gsize bufsz = 0;
+	const guint8 *buf;
+	gsize offset;
+	guint8 image_end_magic[TAP_TOUCH_MAGIC_TAG_LEN];
+	const guint16 ap_polynomial = 0x8408;
+	guint16 ap_crc = 0;
+	guint32 df_checksum = 0;
+	guint16 fw_ic_name;
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(FuStructLogitechTapTouch) st = fu_struct_logitech_tap_touch_new();
+	guint write_retries = 0;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 0, "initialize");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_READ, 0, "metadata");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 3, "erase");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 3, "write-df-blocks");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 94, "write-ap-blocks");
+
+	/* small image size, load entire image, scattered magic tag/metadata */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* validate firmware file size, typical < 75k */
+	buf = g_bytes_get_data(fw, &bufsz);
+	if (bufsz > TAP_TOUCH_MAX_FW_FILE_SIZE) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_DATA,
+			    "failed to validate firmware file size:%" G_GSIZE_FORMAT,
+			    bufsz);
+		return FALSE;
+	}
+
+	/* enable/disable TDE mode */
+	locker =
+	    fu_device_locker_new_full(FU_DEVICE(self),
+				      (FuDeviceLockerFunc)fu_logitech_tap_touch_device_enable_tde,
+				      (FuDeviceLockerFunc)fu_logitech_tap_touch_device_disable_tde,
+				      error);
+	if (locker == NULL)
+		return FALSE;
+	/* recommended to retry at least once */
+	do {
+		fu_progress_step_done(progress);
```

Comment:
I think all of this belongs in a `fu_device_retry` helper --it'll also unindent the code a bit too.

Reply:
Fixed